File: plugins/weather.py
import re
from flask import Flask
from flask import Response
from datetime import datetime
from mattermost_bot.bot import listen_to
from mattermost_bot.bot import respond_to
import configparser 
import requests
import time
import json

@respond_to('weather$', re.IGNORECASE)
@respond_to('weather ([A-z_-]+$)', re.IGNORECASE)
def weather(message, location=None):
    text = get_weather(location)
    message.send(text)

@respond_to('weather ([0-9]*$)', re.IGNORECASE)
def weather_id(message, city_id=None):
    text = get_weather_id(city_id)
    message.send(text)

weather.__doc__ = "Get city weather"

app = Flask(__name__)
url = "http://api.openweathermap.org/data/2.5/forecast/daily?q={location}&APPID={key}&cnt=5".format
url_id = "http://api.openweathermap.org/data/2.5/forecast/daily?id={city_id}&APPID={key}&cnt=5".format

def init():
    """Initializes variables"""
    global url
    global url_id
    global config_api_key
    Config = configparser.ConfigParser()
    Config.read("configuration/config")
    Config.sections()

    config_api_key = Config.get('OpenWeather', 'api_key')

init()

def get_weather_id(city_id=None):
    request_url = url_id(city_id=city_id, key=config_api_key)
    r = requests.get(request_url)
    if r.json()['cod'] == "502":
        return "fuck you"
    text = build_response_text(r.json(), r.json()['city']['name'])
    return text

def get_weather(location=None):
    # weird bug where I don't know why len(location) is 1 when no location is passed...
    if (location is not None):
        location.strip(' \t\n\r')
        valid = re.match('^[\w-]+$', location) is not None
        if (valid):
             request_url = url(location=location, key=config_api_key)
        else:
            return 'Bin voyons donc ca existe pas ' + location + ' comme ville'
    else:
        request_url = url(location='Montreal', key=config_api_key)
        location='Montreal'
    request_url = url(location=location, key=config_api_key)
    r = requests.get(request_url)
    text = build_response_text(r.json(), location)
    return text

def get_embedded_icon_url(icon_code, desc):
    """Returns a formatted markdown line to show the weather icon"""
    return '![desc](http://openweathermap.org/img/w/{code}.png "{desc}")'.format(code=icon_code, desc=desc)

def get_day_weather_line(day):
    """Return a markdown formatted line for a weather day"""
    day_weekday = datetime.fromtimestamp(day['dt']).strftime("%A")
    day_month = datetime.fromtimestamp(day['dt']).strftime("%b")
    day_day = datetime.fromtimestamp(day['dt']).strftime("%d")
    day_info_date = """{weekday}, {month}. {day_number}""".ljust(25).format(weekday=day_weekday, month=day_month, day_number=day_day)
    day_desc = day['weather'][0]['description']
    day_desc.ljust(50)
    day_temp_high = int(day['temp']['max'] - 273.15)
    day_temp_low = int(day['temp']['min'] - 273.15)
    day_icon = get_embedded_icon_url(day['weather'][0]['icon'], day_desc)
    return "| {day_info_date_param} | {desc_param}  {day_icon_param} | {day_temp_high_param} °C | {day_temp_low_param} °C  |\n".format(day_info_date_param=day_info_date, desc_param=day_desc, day_icon_param=day_icon, day_temp_high_param=day_temp_high, day_temp_low_param=day_temp_low)

def build_response_text(data, location):
    """Post formatted data to mattermost instance"""
    days = []
    payload_text = """
### Weather in {location} for the next few days

| Day | Description | High | Low |
|:---------------------------|:------------------------------------|:--------|:--------|\n""".format(location=location)
    for day in data['list']:
        payload_text += get_day_weather_line(day)

    return payload_text

File: pkg/commands/weather.go
package commands

import (
	"fmt"
	"strconv"
	"time"

	owm "github.com/briandowns/openweathermap"
)

type Weather struct {
	apiKey   string
	current  *owm.CurrentWeatherData
	forecast *owm.ForecastWeatherData
}

func NewWeatherClient(apiKey string) (weather Weather, err error) {
	weather.apiKey = apiKey
	c, err := owm.NewCurrent("C", "en", apiKey)
	if err != nil {
		return
	}
	f, err := owm.NewForecast("16", "C", "en", apiKey)
	if err != nil {
		return
	}
	weather.current = c
	weather.forecast = f
	return
}

func (w Weather) GetCurrentWeather(location string) (message string, err error) {
	err = w.current.CurrentByName(location)
	if err != nil {
		return
	}

	temp := strconv.FormatFloat(w.current.Main.Temp, 'f', 0, 64)
	feel := strconv.FormatFloat(w.current.Main.FeelsLike, 'f', 0, 64)
	wind := strconv.FormatFloat(w.current.Wind.Speed, 'f', 1, 64)
	// dayT := strconv.FormatFloat(day.Temp.Day, 'f', 0, 64)

	message = fmt.Sprintf(`### Current weather in %s

| Description | Temperature | Feels Like | Humidity | Wind |
|:---------------------------|:------------------------------------|:--------|:--------|:--------|:--------|
| %s | %s °C | %s °C | %d%% | %s km/h |`,
		w.current.Name,
		getDescriptionWithIcon(w.current.Weather[0].Icon, w.current.Weather[0].Description),
		temp,
		feel,
		w.current.Main.Humidity,
		wind)

	return
}

func (w Weather) GetWeather(location string) (message string, err error) {
	w.forecast.DailyByName(location, 5)

	message = fmt.Sprintf(`### Weather in %s for the next few days

| Day | Description | High | Low | Humidity | Day |
|:---------------------------|:------------------------------------|:--------|:--------|:--------|:--------|`, location)

	forecast := w.forecast.ForecastWeatherJson.(*owm.Forecast16WeatherData)
	for _, day := range forecast.List {
		text := getWeatherLine(day)
		message += text
	}
	return message, nil
}

// Return a markdown formatted line for a weather day
func getWeatherLine(day owm.Forecast16WeatherList) string {

	time := time.Unix(int64(day.Dt), 0)

	min := strconv.FormatFloat(day.Temp.Min, 'f', 0, 64)
	max := strconv.FormatFloat(day.Temp.Max, 'f', 0, 64)
	dayT := strconv.FormatFloat(day.Temp.Day, 'f', 0, 64)

	return fmt.Sprintf(`
| %s, %s. %d | %s | %s °C | %s °C | %d%% | %s °C |`,
		time.Weekday().String(),
		time.Month().String(),
		time.Day(),
		getDescriptionWithIcon(day.Weather[0].Icon, day.Weather[0].Description),
		max,
		min,
		day.Humidity,
		dayT)
}

func getDescriptionWithIcon(iconCode string, description string) string {
	return fmt.Sprintf("%s ![%s](http://openweathermap.org/img/w/%s.png \"%s\")", description, description, iconCode, description)
}
